### PR TITLE
Add access to field decay values in SimulationData

### PR DESCRIPTION
**Pull request by mardolph-ruco**
Relates to #2550 

Changes:

- Parse log to access step and field decay information as property attribute `sim_data.field_decay`, return as `TimeDataArray`
- adapt `final_decay_value`: use `field_decay` field, take last entry

Example usage:
```
field_decay = sim_data.field_decay
field_decay.plot()
plt.xlabel("Time Step")
plt.ylabel("Field Decay")
plt.yscale("log")
plt.show()
```

image

## Greptile Summary

This PR adds programmatic access to field decay values from simulation logs in the Tidy3D framework. Previously, users could only visually inspect field decay information in log outputs, but had no way to extract numerical values for analysis or plotting.

The implementation introduces a new `field_decay` property to the `SimulationData` class that parses the simulation log using regex patterns to extract time step and field decay information. The regex pattern matches log entries like "- Time step 827 / time 4.13e-14s (4 % done), field decay: 0.110e+00" to extract both the step number and decay value. This data is returned as a `TimeDataArray` with time steps as coordinates, making it consistent with other time-series data in the framework.

The existing `final_decay_value` method is refactored to use this new `field_decay` property, maintaining backward compatibility while reducing code duplication. The implementation includes proper error handling for cases where the log is None (raises `DataError`) or when no field decay values are found (returns a warning and defaults to 1.0 for `final_decay_value`).

This change integrates well with the existing codebase by following established patterns for data access properties and using the same `TimeDataArray` structure used elsewhere in the framework. The functionality enables users to create convergence plots and programmatically assess simulation quality, which is valuable for validation and optimization workflows.

## Confidence score: 4/5

- This PR is generally safe to merge with minimal risk of introducing bugs or breaking existing functionality.
- The implementation follows good practices with proper error handling, maintains backward compatibility, and includes comprehensive tests covering normal operation and edge cases.
- The `tidy3d/components/data/sim_data.py` file needs careful review to ensure the regex pattern correctly handles all possible log formats and that the error handling is robust.

### CODE REVIEW

File: tidy3d/components/data/sim_data.py [hunk 1]
```diff
@@ -932,20 +933,28 @@ class SimulationData(AbstractYeeGridSimulationData):
...
+        decays = [float(m[1]) for m in matches]
+        return TimeDataArray(decays, coords={"t": steps})
+
+    @property
+    def final_decay_value(self) -> float:
+        """Returns value of the field decay at the final time step."""
+        field_decay = self.field_decay
+        if len(field_decay) == 0:
+            log.warning("No field decay values found, using 1.0 as final decay value.")
+            return 1.0
```
Comment (yaugenst-flex) [inline tidy3d/components/data/sim_data.py:956]: Does `1.0` make sense in this case or should it rather be `None`? No strong opinion either way but wondering why you decided to go with this?

File: tidy3d/components/data/sim_data.py [hunk 2]
```diff
@@ -932,20 +933,28 @@ class SimulationData(AbstractYeeGridSimulationData):
     )
 
     @property
```
Comment (yaugenst-flex) [inline tidy3d/components/data/sim_data.py]: This should probably be a cached property. Note we use `td.components.base.cached_property`.

**Review by yaugenst-flex — COMMENTED** [review]
Thanks @mardolph-ruco this looks great! Really have only some very minor nitpicks.

File: tidy3d/components/data/sim_data.py (hunk 1 above)
Comment (mardolph-ruco) [reply tidy3d/components/data/sim_data.py:956]: Had the same thought, chose 1.0 as this was the behavior before.
Actually, 1.0 could be misleading here https://github.com/flexcompute/tidy3d/blob/b58ae5f56ad488d5bebaf0840ee96c5f90cb0978/tidy3d/web/api/tidy3d_stub.py#L266 if the `shutoff_value` was reached before the first progress log. Not sure, maybe change to None?
Alternatively, the log could be checked for "Field decay smaller than shutoff factor, exiting solver."

File: tidy3d/components/data/sim_data.py (hunk 2 above)
Comment (mardolph-ruco) [reply tidy3d/components/data/sim_data.py]: makes sense!

**Review by tylerflex — APPROVED** [review]
Looks great to me, thanks for the attention to detail!

**Comment by tylerflex** [comment]
@tomflexcompute might be good to add an FAQ mention to let users know they can use

```py
field_decay = sim_data.field_decay
field_decay.plot()
plt.xlabel("Time Step")
plt.ylabel("Field Decay")
plt.yscale("log")
plt.show()
```

from here on

File: tidy3d/components/data/sim_data.py (hunk 1 above)
Comment (tylerflex) [reply tidy3d/components/data/sim_data.py:956]: my understanding is that the field decay signifies the ratio of the current average field value to the maximum value seen so far. So even if it is not logged, I think 1.0 is a reasonable assumption. This is probably why we went with it originally.

I'd be ok with just leaving this as default and reassessing this later if we ever want to improve how we do the field decay logging, eg with a backend storage and potentially checking it more often on the backend.

Comment (yaugenst-flex) [reply tidy3d/components/data/sim_data.py:956]: ~~Yes exactly. I think we should just be explicit and return `None` if there is no final decay value. Generally this would likely point to a problem, but I would say that's the caller's concern then 😄~~

For the sake of maintaining backward compatibility let's just stick to the old behavior. We should revisit this at some point though.

**Review by yaugenst-flex — APPROVED** [review]

**Comment by yaugenst-flex** [comment]
@mardolph-ruco thanks this looks ready to go! Could you squash your commits and rebase the branch, and we can merge once CI passes!

**Review by tomflexcompute — APPROVED** [review]
This is great to have!
